crates/pessimistic-proof-program/src/main.rs

```diff
@@ -1,14 +1,35 @@
 #![no_main]
 
-use pessimistic_proof::{certificate::Certificate, generate_leaf_proof, LocalNetworkState};
+use pessimistic_proof::{batch_header::BatchHeader, generate_leaf_proof, LocalNetworkState};
 
 sp1_zkvm::entrypoint!(main);
 
 pub fn main() {
     let initial_state = sp1_zkvm::io::read::<LocalNetworkState>();
-    let certificate = sp1_zkvm::io::read::<Certificate>();
+    let batch_header = sp1_zkvm::io::read::<BatchHeader>();
```

Comment:
>This PR is in-progress and designed to give an idea of the changes required **to move from global to local pessimistic proof generation**

As I understand it, the API of the PP in this PR remains the same as it is today in `main`. The current PP in the `main` branch is already for the local pessimistic proof (and not for the previous global proof approach): it takes one `LocalNetworkState` (the local state of one CDK) and one `Certificate` (aka `BatchHeader`) and ensures that the `Certificate` can be applied on top of the `LocalNetworkState` of the sending network.

The proto on the nullifier that you added in the `pessimistic-proof` will be helpful as we focused on the state management so far. I'll probably cherry-pick some of this once we get on it :ok_hand: 

>This is highly beneficial as it removes the requirement for networks to sync with global state, or for all AggLayer nodes to maintain complete Local Balance Trees for every chain.

I believe that one point of misalignment/confusion might be on where the `LocalNetworkState` should live, and by extension who should compute the proof.

To give back a bit of history on the spec decision, we had basically two options as proposed by Carlos hackmds:

Option 1:
- Each `LocalNetworkState` lives in their own CDK stack, but it'd require:
  - one component (or modif of an existing one) on the CDK stack to store and track the new needed data structures (e.g., nullifier tree)
  - one component (or modif of an existing one) that does the job of interacting with the SP1 infra and behaving accordingly (probably in Rust)

Option 2:
- The AggLayer service host the `LocalNetworkState` for each Network as an **intermediate step** and will handle the interactions with the SP1 prover infra

We went for the option 2 as mentioned in the rustdoc of the `LocalNetworkState` here:
https://github.com/agglayer/agglayer/blob/3f7062a58e21615bae38816758d5b2ea18a2114b/crates/pessimistic-proof/src/local_state.rs#L11-L13
    
The nice thing with the option 2 is that the preliminary development and definition of the state for the pessimistic proof lives just aside of the pessimistic proof itself which gives great flexibility for the first iteration. Once the functionality gets established, it'd totally make sense to split it up properly as currently planned.

Reply:
> As I understand it, the API of the PP in this PR remains the same as it is today in `main`

Yep, it remains the same in that we're taking some data in a batch header and applying it to the local network state. However, the `BatchHeader` and `LocalNetworkState` have changed substantially.

> The proto on the nullifier that you added in the pessimistic-proof will be helpful as we focused on the state management so far. I'll probably cherry-pick some of this once we get on it 👌

I'll finish the impl (mostly with help from Will B as I'll likely be on paternity leave shortly).

> To give back a bit of history on the spec decision, we had basically two options as proposed by Carlos hackmds:

Yep understood, I had initially proposed the fully local version, where proof generation is moved off of the AL in a hackmd. It doesn't matter to me where this info is stored or whether there's a separate service now or in the future. 

My priority with this PR is to get to the point where the API for PP generation matches what we want it to be, both from the perspective of fully local proof generation and preparing for fast interop.